File: Managed Source/Arsenal.ImageMounter.Devio/Server/Interaction/RAMDiskService.vb
Imports System.IO
Imports System.Security.AccessControl
Imports System.Threading
Imports Arsenal.ImageMounter.Devio.Server.Services
Imports Arsenal.ImageMounter.IO
Imports DiscUtils
Imports DiscUtils.Partitions
Imports DiscUtils.Raw
Imports DiscUtils.Streams
Imports Buffer = System.Buffer

Namespace Server.Interaction

    Public Class RAMDiskService
        Inherits DevioNoneService

        Public ReadOnly Property Volume As String

        Public ReadOnly Property MountPoint As String

        Public Sub New(Adapter As ScsiAdapter, DiskSize As Long, FormatFileSystem As InitializeFileSystem)
            MyBase.New(DiskSize)

            Try
                StartServiceThreadAndMount(Adapter, 0)

                Using device = OpenDiskDevice(FileAccess.ReadWrite)

                    device.DiskPolicyReadOnly = False

                    ' Initialize partition table

                    device.DiskPolicyOffline = True

                    Dim kernel_geometry = device.Geometry.Value

                    Dim discutils_geometry As New Geometry(
                        device.DiskSize.Value,
                        kernel_geometry.TracksPerCylinder,
                        kernel_geometry.SectorsPerTrack,
                        kernel_geometry.BytesPerSector)

                    Dim disk As New Raw.Disk(device.GetRawDiskStream(), Ownership.None, discutils_geometry)

                    DiscUtilsInteraction.InitializeVirtualDisk(disk, discutils_geometry, NativeFileIO.PARTITION_STYLE.PARTITION_STYLE_MBR, FormatFileSystem, "RAM disk")

                    device.FlushBuffers()

                    device.DiskPolicyOffline = False

                    Do
                        _Volume = NativeFileIO.EnumerateDiskVolumes(device.DevicePath).FirstOrDefault()

                        If _Volume IsNot Nothing Then
                            Exit Do
                        End If

                        device.UpdateProperties()

                        Thread.Sleep(200)

                    Loop

                    Dim mountPoint = NativeFileIO.EnumerateVolumeMountPoints(_Volume).FirstOrDefault()

                    If mountPoint Is Nothing Then

                        Dim driveletter = NativeFileIO.FindFirstFreeDriveLetter()

                        If driveletter <> Nothing Then

                            Dim newMountPoint = $"{driveletter}:\"

                            NativeFileIO.SetVolumeMountPoint(newMountPoint, _Volume)

                            mountPoint = newMountPoint

                        End If

                    End If

                    _MountPoint = mountPoint

                    Return

                End Using

            Catch ex As Exception

                Dispose()

                Throw New Exception("Failed to create RAM disk", ex)

            End Try

        End Sub

    End Class

    Public Enum InitializeFileSystem
        None
        NTFS
        FAT
    End Enum

End Namespace

File: Managed Source/Arsenal.ImageMounter.Forms/RAMDiskService.cs
using Arsenal.ImageMounter.Devio.Server.Services;
using Arsenal.ImageMounter.Extensions;
using Arsenal.ImageMounter.IO.Native;
using DiscUtils;
using DiscUtils.Raw;
using DiscUtils.Streams;
using System;
using System.IO;
using System.Linq;
using System.Threading;
using System.Windows.Forms;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Arsenal.ImageMounter.Devio.Server.Interaction;

public class RAMDiskService : DevioNoneService
{
    public string? Volume { get; }

    public string? MountPoint { get; }

    public RAMDiskService(ScsiAdapter Adapter, long DiskSize, InitializeFileSystem FormatFileSystem)
        : base(DiskSize)
    {

        try
        {
            StartServiceThreadAndMount(Adapter, 0);

            using var device = OpenDiskDevice(FileAccess.ReadWrite)
                ?? throw new NotSupportedException("Cannot open disk device associated with this instance");

            device.DiskPolicyReadOnly = false;

            // Initialize partition table

            device.DiskPolicyOffline = true;

            var kernel_geometry = device.Geometry
                ?? throw new NotSupportedException("Failed to get geometry for RAM disk");

            var kernel_disksize = device.DiskSize
                ?? throw new NotSupportedException("Failed to get disk size for RAM disk");

            var discutils_geometry = new Geometry(kernel_disksize, kernel_geometry.TracksPerCylinder, kernel_geometry.SectorsPerTrack, kernel_geometry.BytesPerSector);

            var disk = new Disk(device.GetRawDiskStream(), Ownership.None, discutils_geometry);

            DiscUtilsInteraction.InitializeVirtualDisk(disk, discutils_geometry, PARTITION_STYLE.MBR, FormatFileSystem, "RAM disk");

            device.FlushBuffers();

            device.DiskPolicyOffline = false;

            do
            {
                Volume = NativeFileIO.EnumerateDiskVolumes(device.DevicePath).FirstOrDefault();

                if (Volume is not null)
                {
                    break;
                }

                device.UpdateProperties();

                Thread.Sleep(200);
            }

            while (true);

            var mountPoint = NativeFileIO.EnumerateVolumeMountPoints(Volume).FirstOrDefault();

            if (string.IsNullOrWhiteSpace(mountPoint))
            {

                var driveletter = NativeFileIO.FindFirstFreeDriveLetter();

                if (driveletter != '\0')
                {

                    var newMountPoint = $@"{driveletter}:\";

                    NativeFileIO.SetVolumeMountPoint(newMountPoint.AsMemory(), Volume.AsMemory());

                    MountPoint = newMountPoint;

                }
            }

            else
            {

                MountPoint = mountPoint.ToString();

            }

            return;
        }

        catch (Exception ex)
        {

            Dispose();

            throw new Exception("Failed to create RAM disk", ex);

        }
    }

    public static RAMDiskService? InteractiveCreate(IWin32Window? _, ScsiAdapter adapter)
    {

        var strsize = Microsoft.VisualBasic.Interaction.InputBox("Enter size in MB", "RAM disk");

        if (strsize is null || string.IsNullOrWhiteSpace(strsize))
        {
            return null;
        }

        if (!long.TryParse(strsize, out var size_mb) || size_mb <= 0L)
        {
            return null;
        }

        var ramdisk = new RAMDiskService(adapter, size_mb << 20, InitializeFileSystem.NTFS);

        if (ramdisk.MountPoint is not null)
        {
            try
            {
                NativeFileIO.BrowseTo(ramdisk.MountPoint);
            }

            catch (Exception ex)
            {
                MessageBox.Show($"Failed to open Explorer window for created RAM disk: {ex.JoinMessages()}");
            }
        }

        return ramdisk;

    }
}
